### PR TITLE
sql: create table can push operations on the table

### LINKED ISSUE #19925: sql: an INSERT following a CREATE TABLE can fail
If a CREATE TABLE is executed via one node at time T followed by an INSERT that is executed on another node at T -delta, the INSERT will fail (). This failure is demonstrated through this test:

```
func TestTxnObeysTableCreationTime(t *testing.T) {
	defer leaktest.AfterTest(t)()
	params, _ := createTestServerParams()
	s, sqlDB, _ := serverutils.StartServer(t, params)
	defer s.Stopper().Stop(context.TODO())

	if _, err := sqlDB.Exec(`
CREATE DATABASE t;
`); err != nil {
		t.Fatal(err)
	}

	tx, err := sqlDB.Begin()
	if err != nil {
		t.Fatal(err)
	}
	
	if _, err := sqlDB.Exec(`
CREATE TABLE t.kv (k CHAR PRIMARY KEY, v CHAR);
`); err != nil {
	t.Fatal(err)
}

	// Insert an entry so that the transaction is guaranteed to be
	// assigned a timestamp.
	if _, err := tx.Exec(`
INSERT INTO t.kv VALUES ('a', 'b');
`); !testutils.IsError(err, "does not exist") {
		t.Fatal(err)
	}

	if err := tx.Rollback(); err != nil {
		t.Fatal(err)
	}
}

```

The correct behavior here is for the INSERT to retry through a ReadWithinUncertaintyIntervalError 
retryable error.

### PR DESCRIPTION
**Pull request by vivekmenezes**
fixes #19925

Release note: None

### CODE REVIEW

**Comment by vivekmenezes** [comment]
The other option is to pass in the txn into AcquireByName() and use it for the table name resolution. That will save one roundtrip to the server.

**Comment by vivekmenezes** [comment]
@andreimatei updated this PR to work with the new txn refresh code. @spencerkimball I'm really unhappy that I've had to use the RefreshTimestamp here. Is it possible for us to make a separate change that updates the OrigTimestamp when the RefreshTimestamp is updated?

**Comment by andreimatei** [comment]
I think "push" is the wrong term to use, but in the commit message and in the test. Say something about "uncertainty", which is different from a push.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/table.go, line 287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21686#-L6x7SK5ERCYFKwM1Sf3:-L6x7SK5ERCYFKwM1Sf4:b-5x3yqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd2b3836d8f6d7fb0e974b180b962dfe78a8ca85/pkg/sql/table.go#L287)):*
> ```Go
> 
> 	origTimestamp := flags.txn.OrigTimestamp()
> 	table, expiration, err := tc.leaseMgr.AcquireByName(ctx, origTimestamp, dbID, tn.Table())
> ```

I think detecting of this uncertainty should be pushed to the lowest level possible; can we move it inside `AcquireByName`?

---

*[pkg/sql/table.go, line 297 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21686#-L6x7MQCBypTMfvnKOj3:-L6x7MQCBypTMfvnKOj4:b1mfkko) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd2b3836d8f6d7fb0e974b180b962dfe78a8ca85/pkg/sql/table.go#L297)):*
> ```Go
> 		phyAccessor := UncachedPhysicalAccessor{}
> 		_, _, errRetry := phyAccessor.GetObjectDesc(tn, uncachedFlags)
> 		if errRetry == nil && flags.txn.Proto().RefreshedTimestamp != (hlc.Timestamp{}) {
> ```

please explain what's going on here with the retries

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21686#-:-L6x6qdv4gUnn_CxW77N:b-ax7caf)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/table.go, line 287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21686#-L6x7SK5ERCYFKwM1Sf3:-L7--L0snSs8WfFHYjb_:b-1m2gts) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd2b3836d8f6d7fb0e974b180b962dfe78a8ca85/pkg/sql/table.go#L287)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think detecting of this uncertainty should be pushed to the lowest level possible; can we move it inside `AcquireByName`?
</blockquote></details>

Yes we can move it down but that will require passing in txn to AcquireByName and using the txn to do the name -> id lookup.  Once that lookup is done I'll have to check that the RefreshTimestamp has been set to see the new pushed timestamp. Should I go ahead and do that? It's still going to look hacky

---

*[pkg/sql/table.go, line 297 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21686#-L6x7MQCBypTMfvnKOj3:-L7--ZVS_9pPGASqyer-:bborpqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd2b3836d8f6d7fb0e974b180b962dfe78a8ca85/pkg/sql/table.go#L297)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please explain what's going on here with the retries
</blockquote></details>

The retry is causing the transaction to be pushed and the RefeshedTimestamp to be set.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21686)*

**Comment by vivekmenezes** [comment]
I can do that. For a bit of historic context here, when I originally made this change the code would return an uncertainty error, but once I rebased over spencer's change I had to start using the RefreshedTimestamp as well as change the unittest to not see an error. i think I'd like to keep RefreshedTimestamp internal to kv if possible

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21686#-:-L7-2XAxpV1SWnt-Ukfe:bdxt14l)*

**Comment by andreimatei** [comment]
note to self: I think we can get the `ReadWithinUncertaintyInterval` errors back if we set `OrigTimestampWasObserved` on whatever txn we want to get those errors.

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21686#-:-L70lAzofIYvIGTc7V5j:b934o83)*

**Comment by vivekmenezes** [comment]
Let me try that

File: pkg/sql/table.go [hunk 1]
```diff
@@ -285,6 +285,20 @@ func (tc *TableCollection) getTableVersion(
...
+	if err == sqlbase.ErrDescriptorNotFound {
+		// The table doesn't exist at the requested timestamp.
+		// Read the table descriptor directly to check if the descriptor
+		// exists at a future timestamp within the clock uncertainty
+		// interval.
+		uncachedFlags := flags
+		uncachedFlags.avoidCached = true
+		phyAccessor := UncachedPhysicalAccessor{}
+		_, _, errRetry := phyAccessor.GetObjectDesc(tn, uncachedFlags)
+		if errRetry == nil && flags.txn.Proto().RefreshedTimestamp != (hlc.Timestamp{}) {
```
Comment (spencerkimball) [inline pkg/sql/table.go]: This seems pretty iffy. Let's discuss in person tomorrow – there must be a better way to do this than breaking the abstraction pretty thoroughly. Code at this level should not be peering into the transaction proto to make decisions.

Comment (andreimatei) [reply pkg/sql/table.go]: As I was telling Vivek earlier, I think the real problem here is that `leaseMgr.AcquireByName()` basically swallows the uncertainty of whatever read it does and lets higher layers fend for themselves. I'll attempt to do something about that as soon as I can, unless Vivek does something else first.

**Comment by vivekmenezes** [comment]
@andreimatei this PR is no more hacky. If you think this is mergeable I'll do the same for getTableVersionByID() and add a test for that. That method is used for FKs

**Comment by andreimatei** [comment]
The leasing mechanism is supposed to compensate for the lack of the timestamp cache... I'm not sure how that ties into the uncertainty discussion... Mind explaining more?

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21686#-:-L75LjeGayoaLDyW_vFC:b-27cyc9)*

**Comment by vivekmenezes** [comment]
@andreimatei I'm trying out another idea which concerns using the transaction to get cached descriptors that are not leased descriptors. I'll give that a try.

**Pull request closed without merging**